CARDS-2526: New form pagination variant: navigable

marta- opened the pull request:
DRAFT because:
* it was only tested on one paginated form as admin; no testing was done as a patient user
* the QuestionnairePreview might also need to be updated (not tested)

**Specs:** : https://phenotips.atlassian.net/browse/CARDS-2526

**Known issues:**
* sometimes `requireCompletion` doesn't work (also happens in `dev` - not introduced in this PR)
* Phone Call Follow-Up - sometimes advancing past the first page requires two attempts (also happens in `dev` - not introduced in this PR)

**To test:**
* Choose a paginated questionnaire or create a new one. Examples of existing paginated questionnaires:
  * `test` : Pagination Test
  * `heracles`: Phone Call Follow-Up
  * `prems`: CPESIC, Rehab
  * `proms`: Smoking cessation
  * `sparc`: Clinician Assessment (the sparc project + navigable pagination are available in branch `sparc-demo-2`
* Set the questionnaire's pagination variant to `navigable`
![image](https://github.com/data-team-uhn/cards/assets/651980/3271c8a0-452d-43d0-976a-9dba54851739)
* Test  the questionnaire preview
* Create a form for the questionnaire
* Test the form with different screen widths
* Test forms with title-less section pages, and with pages made of questions not wrapped in a section

marta- committed:
CARDS-2526: New form pagination variant: navigable

sashaandjic commented:
Not sure if I did something wrong to cause this:sasha@SASHA:~/CARDS/CARDS-2526/cards$ startsparc 7777
CARDS_VERSION 0.9.25-SNAPSHOT
[INFO]
[INFO] Apache Sling Application Launcher
[INFO] ---------------------------------
[INFO] Initializing...
[INFO] Artifact Repositories: [file:///home/sasha/CARDS/CARDS-2526/cards/.mvnrepo, file:///home/sasha/.m2/repository, https://nexus.phenotips.org/nexus/content/groups/public, https://repo.maven.apache.org/maven2, https://repository.apache.org/content/groups/snapshots]
[INFO] Assembling final feature model...
[ERROR] Error while assembling launcher: Artifact mvn:io.uhndata.cards/cards4sparc/0.9.25-SNAPSHOT/slingosgifeature not found in any repository.
java.io.IOException: Artifact mvn:io.uhndata.cards/cards4sparc/0.9.25-SNAPSHOT/slingosgifeature not found in any repository.
        at org.apache.sling.feature.io.artifacts.ArtifactManager.getArtifactHandler(ArtifactManager.java:292)
        at org.apache.sling.feature.launcher.impl.FeatureProcessor.createApplication(FeatureProcessor.java:143)
        at org.apache.sling.feature.launcher.impl.Bootstrap.assemble(Bootstrap.java:219)
        at org.apache.sling.feature.launcher.impl.Bootstrap.run(Bootstrap.java:135)
        at org.apache.sling.feature.launcher.impl.Main.main(Main.java:200)
******************************************************
*                                                    *
*   The CARDS Java process has failed at port 7777   *
*                                                    *
******************************************************
sasha@SASHA:~/CARDS/CARDS-2526/cards$

marta- commented:
@sashaandjic `CARDS_VERSION 0.9.25-SNAPSHOT` hints that you're not on the `sparc-demo-2` branch but on the `CARDS-2526` branch which does not have the `sparc` project.

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -138,25 +159,36 @@ function FormPagination (props) {
+      setActivePage(nextPage);
+      onPageChange?.();
+      // Call the onDone() if we're on the last page
+      if (nextPage === lastValidPage()) {
```
This triggers when advancing to the last page, not when saving the last page

marta- replied:
Regression fixed in 426482a .

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -654,6 +654,12 @@ function Form (props) {
+               Object.values(data)
+                 .filter(e => Object.values(e).find(val => ENTRY_TYPES.includes(val["jcr:primaryType"])))
+                 .filter(e => keys.includes((e.section || e.question)?.["@name"] || ""))
+                 .every(p => !hasWarningFlags(p))
```
This seems to check a stale version of status flags from when the form was opened and does not reflect any changes to answers since then. To see this issue you can use the `--test` `Pagination test` questionnaire with a couple changes:
- Edit to `navigation` style pagination
- Edit the `diagnosis` page to be conditional on only `alive with disease`, not `alive with disease` and `dod`. This is likely required due to a conditional refactor unrelated to this branch

Reproduction steps:
1. Create a new Pagination test form. On creation, the following status flags are set:
- Page 1 is flagged as `INCOMPLETE, DRAFT` due to missing answers
- Page 2 has no flags as the conditional to enable it is not satisfied
2. Complete page 1, being sure to select `Alive with disease` to enable page 2.
3. Do not fill out any answers on page 2
4. Proceed to Page 3. Observe that Page 1 is incorrectly flagged with a warning and Page 2 is incorrectly marked as complete
![image](https://github.com/data-team-uhn/cards/assets/20056751/f7962ad1-5836-43a6-b7f8-ac78679ab40d)
5. Exit the form back to dashboard and begin editing the form again.
6. Proceed to Page 3 and observe that the flags on Page 1 and 2 are correct
![image](https://github.com/data-team-uhn/cards/assets/20056751/e16a0d4a-cb4b-4e45-a74e-d6dd99278b35)
7. Complete Page 2 and return to Page 3. Observe that Page 2 did not update to be complete
![image](https://github.com/data-team-uhn/cards/assets/20056751/3fa133e3-bf91-4ff5-be30-6e85747225cc)

acrowthe replied:
~~This comment is out of date, I figured out more details but forgot to save my changes. Editing...~~
Done, comment is up to date

marta- replied:
Addressed in 43995ce .

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPageNavigation.jsx:
```diff
@@ -0,0 +1,158 @@
+        disablePortal
+        anchorEl={pageSelectorAnchorEl}
+        open={!!pageSelectorAnchorEl}
+        onClose={() => setPageSelectorAnchorEl(null)}
```
```suggestion
        onClose={() => setPageSelectorAnchorEl(null)}
        disableScrollLock={true}
```
Having scrolling disabled when the menu is open feels awkward to me

marta- replied:
Done in 426482a .

acrowthe commented:
A couple more notes:
1. It's really nice that the dropdown variant of the page menu always reflects the pages that are enabled, it would be nice if the bubbles could automatically update too although this is a pre-existing issue with other forms of pagination
![image](https://github.com/data-team-uhn/cards/assets/20056751/89d078ed-1b69-4259-83dc-98695fe1e717)

2. If you open the menu version of navigation, make the page wider to change to the bubble version and then shrink the page to the menu version again the position incorrectly snaps to the top left of the page. It's an edge case issue, may not be worth addressing
![image](https://github.com/data-team-uhn/cards/assets/20056751/b51ae8c1-48a9-4ebd-9ae9-702ee3315343)

marta- committed:
CARDS-2526: New form pagination variant: navigable
Addressing review comments:
* Regression fix in form pagination
* Better display of page list on narrow screens

marta- committed:
CARDS-2526: New form pagination variant: navigable
Addressing review comments: re-fetch the form data to display the correct page status flags in navigable pagination (also minor refactoring)

marta- committed:
CARDS-2526: New form pagination variant: navigable
Addressing review comments: remove page list wrongly placed on narrow screens after resize

marta- commented:
> 2\. If you open the menu version of navigation, make the page wider to change to the bubble version and then shrink the page to the menu version again the position incorrectly snaps to the top left of the page.

Addressed in db44242 .

marta- commented:
> 1. It's really nice that the dropdown variant of the page menu always reflects the pages that are enabled, it would be nice if the bubbles could automatically update too although this is a pre-existing issue with other forms of pagination

I strongly agree, however since this isn't currently supported I propose to address it in a future iteration.

acrowthe reviewed (approved):
Looks good. Tested using Pagination test form and found no issues there

sdumitriu committed:
CARDS-2526: New form pagination variant: navigable
Don't use the node name for the page title, use a generic "Page N" instead

sashaandjic reviewed (commented):
tested
